### PR TITLE
[AARCH-25772,cat] Changes to the ordering requirements of DMB LD and DSB LD

### PR DESCRIPTION
**Pull request by artkhyzha**
This PR relaxes the architectural intent so that DMB LD and DSB LD provide order from Implicit Tag Memory Read Effects only if those are generated by instructions that (1) also generate Explicit Memory Read Effects, and (2) are not atomic instructions that use a zero register as their destination register.

In the case of DSB LD, implementing the relaxation to the architectural intent means correcting the ETS2 clause that previously omitted Implicit Tag Memory Reads. Hence, this specific case actually shows up as a strengthening in logs.

The following is the output of verdicts from `herd7` for the tests in this PR. Some of them illustrate the changes, others illustrate corner cases that are not affected.

```
                                          |Kind| Before After 
-------------------------------------------------------------
-------------------------------------------------------------
ETS2.MP-I2V+dmb.stpt+dsb.ld               |---| Allow  Forbid
-------------------------------------------------------------
LB-noret-tagfault+dmb.ldpp+dmb.ldpt       |---| Forbid Allow 
-------------------------------------------------------------
LB-synctagfault+dmb.ldpp+dmb.ldpt         |---| Forbid Forbid
-------------------------------------------------------------
LB-tagfault+dmb.ldpp+dmb.ldpt             |---| Forbid Forbid
-------------------------------------------------------------
LB-tagfault-casfail+dmb.ld+dmb.ldpt       |---| Forbid Forbid
-------------------------------------------------------------
LB-tagfault-casok+dmb.ld+dmb.ldpt         |---| Forbid Forbid
-------------------------------------------------------------
MP-noret+dmb.ldpt+dmb.ldpp                |---| Forbid Allow 
-------------------------------------------------------------
MP-synctagfault+dmb.ldpt+dmb.ldpp         |---| Allow  Allow 
-------------------------------------------------------------
R+dmb.ldpt+dmb.ldpp                       |---| Forbid Allow 
-------------------------------------------------------------
R-noret-synctagfault+dmb.ldpt+dmb.ldpp    |---| Forbid Forbid
-------------------------------------------------------------
WWRW-asynctagfault+dmb.ldpp+dmb.ldpt      |---| Forbid Allow 
-------------------------------------------------------------
WWRW-noret-asynctagfault+dmb.ldpp+dmb.ldpt|---| Forbid Allow
```

P.S. The above table is produced via commands:
```
herd7 tests/*.litmus > Before
herd7 tests/*.litmus -I libdir > After
mcompare7 -show X Before After
```

### CODE REVIEW

**Comment by artkhyzha** [comment]
~Currently fails `make test`. This is due to the test for `miaou7` introduced in [this commit](https://github.com/herd/herdtools7/commit/df0386f3ddb168bbf2c42f0ee97b37752f4b767b). For the PR to pass that test, it needs improvements to `miaou7` for handling the `domain` function in cat.~

Fixed. A different change implemented to handle "domain". Also, the cat file is edited slightly to improve the output.

File: herd/libdir/aarch64hwreqs.cat [hunk 1]
```diff
@@ -67,8 +67,8 @@ let f-ib =
 let DSB-ob =
 [M | DC.CVAU | IC | TLBI]; po; [dsb.full]; po; [~(Imp & TTD & M | Imp & Instr & R)]
  | (if "ETS2" || "ETS3" then [M | DC.CVAU | IC | TLBI]; po; [dsb.full]; po; [Imp & TTD & M] else 0)
- | [(Exp & R) \ NoRet | Imp & Tag & R]; po; [dsb.ld]; po; [~(Imp & TTD & M | Imp & Instr & R)]
- | (if "ETS2" || "ETS3" then [(Exp & R) \ NoRet]; po; [dsb.ld]; po; [Imp & TTD & M] else 0)
+ | [(Exp & R) \ NoRet | domain([Imp & Tag & R]; same-instance; [(Exp & R) \ NoRet])]; po; [dsb.ld]; po; [~(Imp & TTD & M | Imp & Instr & R)]
```
Comment (relokin) [inline herd/libdir/aarch64hwreqs.cat:70]: have we consider the case where E1 is generated by a LDR instruction that faults?

File: catalogue/aarch64-oban/aarch-25772/libdir/aarch64.cat
Comment (relokin) [inline catalogue/aarch64-oban/aarch-25772/libdir/aarch64.cat:1]: Can you please also add the catdefinitions.tex here?

I assume you didn't have to make any changes to catdefinitions.tex?

File: herd/libdir/aarch64hwreqs.cat (hunk 1 above)
Comment (artkhyzha) [reply herd/libdir/aarch64hwreqs.cat:70]: Yes and no. It is quite hard to come up with a relevant test. In simple MP shapes, this Imp Tag R would be IFB-ordered-before everything that follows the exception entry in program order. This renders DMB LD irrelevant.

So in principle it is interesting to ponder whether the phrasing for the changes to DMB LD needs to be adjusted to account for the faulting case, but it could be good to invent a test for it.

File: tools/miaou.ml [hunk 2]
```diff
@@ -564,6 +564,9 @@ and cons_seqs (fs:exp list) (es:exp list) =
            List { l with items = List.rev items; }
         | _ as i -> i
         end
+      | App (_,Var (_,"domain"), rel) ->
+        let e3 = Next.next () in
+        tr_rel e1 e3 rel
```
Comment (TiberiuBucur) [inline tools/miaou.ml:569]: I believe @relokin had a legit implementation for domain in [#1864](https://github.com/herd/herdtools7/pull/1864), which got dropped before the PR was merged. Perhaps we could add that here.

**Review by TiberiuBucur — COMMENTED** [review]
Overall LGTM, but I see @relokin 's point about faulting LDRs. Will we target that in this PR or are we going to let it progress for now?

File: catalogue/aarch64-oban/aarch-25772/tests/ETS2.MP-I2V+dmb.stpt+dsb.ld.litmus
Comment (TiberiuBucur) [inline catalogue/aarch64-oban/aarch-25772/tests/ETS2.MP-I2V+dmb.stpt+dsb.ld.litmus:1]: Perhaps we can add a kinds file in the tests directory?

File: herd/libdir/aarch64hwreqs.cat (hunk 1 above)
Comment (artkhyzha) [reply herd/libdir/aarch64hwreqs.cat:70]: Marking this as resolved, since no further cat-file changes are planned.

File: catalogue/aarch64-oban/aarch-25772/tests/ETS2.MP-I2V+dmb.stpt+dsb.ld.litmus
Comment (artkhyzha) [reply catalogue/aarch64-oban/aarch-25772/tests/ETS2.MP-I2V+dmb.stpt+dsb.ld.litmus:1]: Done.

File: catalogue/aarch64-oban/aarch-25772/libdir/aarch64.cat
Comment (artkhyzha) [reply catalogue/aarch64-oban/aarch-25772/libdir/aarch64.cat:1]: Done. Indeed no changes.

File: tools/miaou.ml (hunk 2 above)
Comment (artkhyzha) [reply tools/miaou.ml:585]: Is the relevant commit still available, @relokin? Happy to include a better implementation for "domain", if there is one.

File: catalogue/aarch64-oban/aarch-25772/libdir/catdefinitions.tex
Comment (TiberiuBucur) [inline catalogue/aarch64-oban/aarch-25772/libdir/catdefinitions.tex]: This leaves out some commands that were added with the M.d release (present on master). Since this will come afterwards, should we include them as well?

File: catalogue/aarch64-oban/aarch-25772/kinds.txt [hunk 3]
```diff
@@ -0,0 +1,12 @@
+ETS2.MP-I2V+dmb.stpt+dsb.ld                     Forbid
+LB-noret-tagfault+dmb.ldpp+dmb.ldpt             Allow
+LB-synctagfault+dmb.ldpp+dmb.ldpt               Forbid
+LB-tagfault+dmb.ldpp+dmb.ldpt                   Forbid
+LB-tagfault-casfail+dmb.ld+dmb.ldpt             Forbid
+LB-tagfault-casok+dmb.ld+dmb.ldpt               Forbid
+MP-noret+dmb.ldpt+dmb.ldpp                      Allow
+MP-synctagfault+dmb.ldpt+dmb.ldpp               Allow
+R+dmb.ldpt+dmb.ldpp                             Allow
+R-noret-synctagfault+dmb.ldpt+dmb.ldpp          Forbid
```
Comment (TiberiuBucur) [inline catalogue/aarch64-oban/aarch-25772/kinds.txt]: The test name is actually `R-noret-tagfault+dmb.ldpt+dmb.ldpp.litmus`.
